### DISCUSSION TITLE
Better boolean attribute behavior

### DISCUSSION THREAD

**Discussion by jfbrennan** [vuejs/rfcs#409] [RFC Discussions]
Ok, I'm trying this again. Apparently opening a "Feature Request" through Vue's Issue Helper is not where I'm supposed to go for feature requests. I was promptly sent here by @posva who is clearly annoyed by this topic. I'm sorry, I don't mean to irritate, but you never said to open an RFC, you told me to go read the old ones.

There's a genuine need for discussion here and from my pov it's being overly simplified, dismissed, and then silenced by the Vue team. For example, "you are the only person complaining about the improvement of writing :aria-value="boleanValue" instead of :aria-value="booleanValue.toString()"" is a very narrow view of the problem space and is resulting in hasty decision-making that doesn't address the nuanced needs of developers.

Below was copied and re-markdown'ed from https://github.com/vuejs/vue/issues/12342 because I was locked out.

---

## What problem does this feature solve?

**I understand there is already bias against this topic, but I'd ask for a chance to revisit it in more detail.**

After the necessary removal of proprietary **enumerated** attributes behavior, the **boolean** attributes behavior has been made less intuitive, more difficult, and error-prone.

**Less intuitive** because HTML attributes take string values, so it was widely understood a boolean value of `false` would result in no attribute at all, e.g. `<div :attr="false">` would result in `<div>`. This was ideal because the code for controlling these attributes naturally produced boolean values, i.e. it Just Worked. Attributes that need strings "true" and "false" were easily provided with `bool.toString()`, which made sense because HTML attributes only take strings. More on these attributes below.

Also, Vue now matches native `setAttribute('attr', false)` which results in `attr="false"`, so you might expect Vue to give `null` equal treatment, e.g. `setAttribute('attr', null)` results in `attr="null"`, but it doesn't.

**More difficult** because developers now have to work against JavaScript to ensure their data, template expressions, methods, and/or computed properties result in an unnatural `null` value. It may seem as simple as `:attr="bool || null"` but any computed or method with logic branches the dev now has to go the extra mile to ensure `null` and not `false` is going to be the final output of the branching. Also, this logic `:attr="bool || null"` just makes my brain go Huh? If false give me a null? Never written an expression like that before, would hate to have to scatter that all over my templates.

**More error-prone** because the odds of a value being a boolean are very high and the ubiquitous boolean is now a liability in templates. One way to mitigate this new risk is to make all things that could be `false` be `null` instead. This avoids the possibility of another dev using something that isn't boolean attribute safe. A terrible pattern, but if it means reducing risk devs may go that route. Gross.

Okay, on to attributes that take `"true"` and `"false"` strings...

It seems like what's happened is Vue has replaced some old special behavior (i.e. enumerated attributes) for new special behavior based on assumptions about the importance of a relatively small number of attributes that only take the strings "true" or "false". In short, "true" or "false" was like a decoy that resulted in introducing odd behavior.

Let me explain with some examples:

- Sure, attributes like `aria-required` need a string value of "true" or "false", however...
- Attributes like `contenteditable` are likely given string values _by the developer_ because "true" and "false" are just two of several possible strings the developer can set on that attribute. Devs can write code like `:contenteditable="plainText ? 'plaintext-only' : 'false'"`, so it's strange for Vue to go out of its way to cast booleans to strings for the developer to support this case.
- Same story with `aria-invalid`. It supports the strings "grammar", "spelling", "true", or "false". The developer will supply one of the four strings, not one of two strings or a boolean. Yes, someone could write the latter, but you see where I'm going.
- Same again for `aria-haspopup`, which supports **7 different string values**, including "true" and "false".
- And then there's `aria-expanded` and `aria-pressed`, which support "true", "false", and "undefined" (pressed also supports a fourth string "mixed"). But if I give Vue `undefined` it doesn't cast it to a string like it does for `false` it removes that attribute, which made sense in Vue 2 but not now. Vue requires the developer to supply the string "undefined", but casts `false` for them. Vue does this because it mistakenly believes there's something special about attributes that take "true" or "false" strings. This is not great and it's made even more weird because `setAttribute('attr', undefined)` normally results in `attr="undefined"`.

I think true and false string values can be awkward to reason about at first, so I can emphasize, but developers are now put in an even more awkward situation trying to ensure `false` becomes `null` as well as navigating what Vue will and won't do for them in regards to casting strings for attributes.

**The fix**
Developers are required to provide whatever string their attributes need, including the not special "true" and "false" strings, while Vue reserves the power of `false` (and null or undefined) to remove attributes.

## What does the proposed API look like?
```html
These three values remove an attribute (like removeAttribute('attr')) 
<div :attr="false"> => <div>
<div :attr="null"> => <div>
<div :attr="undefined"> => <div>

True adds an attribute with no value, i.e. a boolean attribute  (using setAttribute('attr', '') not setAttribute('attr', true))
<div :attr="true"> => <div attr>

Everything else (like setAttribute('attr', val))
<div :attr="'true'"> => <div attr="true">
<div :attr="'false'"> => <div attr="false">
<div :attr="'null'"> => <div attr="null">
<div :attr="'undefined'"> => <div attr="undefined">
<div :attr="0"> => <div attr="0">
<div :attr="''"> => <div attr="">
<div :attr="'hi'"> => <div attr="'hi'">
```

**Comment by LinusBorg** [comment]
Hi!

## On being "silenced"

Before I get into the meat of the proposed change, i have to strongly reject your framing of being "silenced" by us.

You were explicitly told to start a discussion in this RFC repo, because significant changes to the framework require and RFC and are usually started with an informal discussion thread here before being formalized (you can read more on the process in this repo's README).

Yes, we locked the issue the discussion started in, because that was a years-old issue in the Vue 2 repository which you seemed to have found and commented on years after the fact. And we closed the feature issue that you opened subsequently because you were asked to start an RFC discussion here. _(By the way, and this is our fault for not properly noting this: vuejs/vue is the Vue 2 repository. Vue 3 lives in vuejs/vue-next at least for a few more weeks.)_

And yes, we made clear that we don't see the severity that you seem to feel about this and aren't very into this. That's not being "silenced", though - that's just our opinion / POV.

## Why an RFC Discussion? 

Because what you are asking is a **big** deal. 

You are asking us to revert a change that went through the RFC process with little objections, has been live in a stable release for more than **13 months**. 

You are asking us to revert a change whose reversal would potentially **break** aria attributes in **every Vue 3 app out there**, and given how challenging it is to automatically test for some a11y stuff, this might even go into production in a lot of projects without being noticed.

So if we accept this proposal, it will be for Vue 4, unless you can give some incredibly convincing argument about how the current is currently actively breaking Vue 3 apps all over the place - which so far, you didn't - and I don't think that's the argument you are trying to make anyway.

## Response to your proposal

### Scope

Since you broadly speak of "boolean attributes", I'd like to preface my response with a bit of a definition of the scope, to the best of my understanding:

*Actual* boolean attributes as per HTML spec (i.e. `input.checked`) are not affected. they work as you expected them to because that's what makes sense for them.

Your issue (*and here I extrapolate from the previous conversation and not so much your points in this proposal*) is with **custom attributes** that are semantically meant to be used as boolean attributes (i.e. because a custom element wants to treat them that way, or because a CSS selector wants to select for `div[expanded]`

And concerning web components, the issue would only be an issue insofar as the custom element implementation doesn't provide a DOM property alongside the attribute - because Vue 3's default behavior is to set a DOM property if it exists, and fall back to an attribute only if that doesn't exist or you [explicitly ask to set it as an attribute](https://v3.vuejs.org/api/directives.html#v-bind).

#### Summary

This problem that you see doesn't really impact Vue developers unless they work with a specific kind of custom element implementation, or a CSS framework relying on custom boolean attributes for selectors - both of which are rare in my personal experience at least.

### About confusion and odd behavior.

#### "Less intuitive"

Yes and no. I agree that, especially if you are used to the Vue 2 behavior, this can be a little confusing at first, provided that you come across a situation where it actually comes to play.

However, that also depends a bit on perspective. The current behavior is simply: 

> _"For any non-boolean attribute (and this includes any custom attributes), we *always* stringify the value, no matter the type: `string | object | array | number | boolean` - you name it, we stringify it."_
> _"The only exception being `null | undefined`, as those don't make much semantic sense as strings in attributes for 99,99999% of cases (while `'true' | 'false'` *do* have semantic meaning in a number of HTML attributes). Therefore, we use those two values to signal a removal."_


#### "More difficult"

Yes. Undoubtedly, `:attr="myBool || null"` a little more hassle and confusing to use than a simple `:attr="myBool"`. No argument there.

#### "More error-prone"

* **Yes** for custom properties, to the extend that you deal with such attributes, which again, in my personal opinion, is rare in a typical Vue project.

* **No** for `aria` attributes, which are hopefully quite common in every Vue application as A11y should be a priority concern.

#### In the list of examples concerning `aria` attributes

You demonstrate that a lot of these aria attributes, as well as i.e. `contenteditable`, accept string values besides `'true'` and `'false'`. However, to me these examplify nicely why stringifying boolean values is actually preferable.

Take for example `aria-haspopup`. in our Javascript, it would be nice to have its value with either a string or an actual boolean because then we can do:

```js
// string (truthy) or actual boolean in JS, nice and intuitive
if (hasPopup) {} else {}
```
```html
<!-- cast to string in HTML, no customization required -->
<div :aria-haspopup="hasPopup">
```

compared to Vue 2 behavior:

```js
// Either do this in your JS:
if (hasPopup !=== 'false') {} else {}
```
```html
<!-- or do this in your HTML -->
<div :aria-haspopup="hasPopup.toString()"
```

So the point is: we basically have to decide which of these two we value more:

* handle aria attributes, contenteditable etc. more intuitively.
* handle custom attributes that are meant to be boolean more intuitively.


Sidenote: What I give you the issue with `aria-expanded` and `'undefined'`. This is the web, there's a weird edge case for _everything_.


### Impact in popular web component libraries

I took a brief(!) look at the two most popular CE libraries (in my perception).

* In StencilJS, you can use `'true' | 'false'` on boolean attributes, it handles them fine.
* In Lit, all public properties are available as DOM properties, so Vue would, by default, set the boolean value to that DOM property, not the attribute, so no stringification would happen. Lit also allows to set a custom value converter, so even for attributes you could parse`'true' | 'false'` into their expected boolean values with like 2 lines of code.
  * This is also in line with best practice recommendations for how to handle properties according to this guide at least: https://developers.google.com/web/fundamentals/web-components/best-practices#aim-to-keep-primitive-data-attributes-and-properties-in-sync,-reflecting-from-property-to-attribute,-and-vice-versa.

I don't know how other CE libraries handle this, but it stands to reason that they offer similar behavior/customization capabilities.

Keep in mind that I'm by no means a CE expert or even regular user, so there might very well be aspects about this that I missed.


### Impact in CSS libraries/frameworks

I'm not personally aware of frameworks that rely on custom boolean attributes (though they surely exist), so I can't really say anything about that.


### Comparison with other Frameworks

So, what do other FE Frameworks do?

* React, Angular, Svelte and Solid work like Vue 3
* Preact works like Vue 2

Given that, if we reverted this change for Vue 3, that would mean that the majority of the apps out there work like the current behavior - which means that custom element libraries as well as CSS libraries are kinda forced to take this into account.

CE Frameworks i.e. do so by casting `'true' | 'false'` strings to the correct boolean values for attributes they expect to be boolean (see StencilJS, for example).

The same goes for CSS Frameworks that might use boolean custom attribute selectors: If they want to be usable without too much hassle in the majority of frontend frameworks, they have to adapt to these frameworks' shared behavior here - regardless of wether or not we revert to Vue 2 behavior according to your proposal.

So in terms of the community as a whole, chances are that most Vue developers won't run into the issues you seem to have ran into, as they would presumably use CE / CSS frameworks and libraries that take into account this behavior across the popular frontend frameworks.

## Summary

* yes, the current behavior is a bit less ergonomic and intuitive when working with custom properties that need to be set as boolean attributes.
  * ...but it's a problem that, from my initial analysis, doesn't affect a large percentage of users.
* At the same time, the current behavior is *more* ergonomic when working with `aria` attributes as shown above.
  ...and those are hopefully *very* common in apps written in Vue 3.

Given that, I don't see how we could justify breaking current user's apps. But maybe we get different user feedback in this discussion, maybe I so far haven'T spoken with the right people that do experience this as a day to day challenge - this is in part what these discussions are for - to get feedback.

All that being said, I understand that this doesn't solve your personal challenge where you seem to be in a situation that *heavily* relies in custom boolean attributes, and I can sympathize with the irritation this is causing you. But we can't introduce breaking changes to the whole community because of the seemingly edge-case-y situation of a few.

**Comment by jfbrennan** [comment]
**Thanks for reading and for thank you for your thorough reply. I honestly thought this was going to closed and locked for a third time.**

>You were explicitly told to start a discussion in this RFC repo

I wasn't, but I assumed "You can also propose a new discussion to change this behavior" meant open a Feature Request issue because that's where a lot of people make suggestions. Didn't know about Discussions. My bad.

>we locked the issue...you seemed to have found and commented on years after the fact

Hey now, I joined the conversation _after_ you **reopened** it. I first commented on an **open** ticket. Don't blame me.

And yes, locking the conversations feature is definitely silencing. Doing that always comes across as rude and disrespectful. People take it as, "I think you need to shut up, so I'll make you shut up." No hard feelings, but come on there's no harm in letting people speak. Btw, comments on closed tickets are important - they can often lead to new information as the community catches up with the authors. Also, no one was breaking any community guidelines so no need to be draconian. 

>unless you can give some incredibly convincing argument about how the current is currently actively breaking Vue 3 apps all over the place

Well, of course Vue 3 projects aren't "actively breaking all over the place" because these breaking changes have to be fixed during the migration lol.

That brings up another issue. It's really disheartening to see the Vue team tell me "you are the only person complaining" when the majority of the community hasn't even been exposed to Vue 3 yet:
<img width="524" alt="Screen Shot 2021-10-29 at 12 05 03 PM" src="https://user-images.githubusercontent.com/1817868/139475133-bcf205c4-d3d6-4c75-b2d4-d8c9e153316e.png">

It's almost 10:1 and that's just one data point on where the community is. Another being all the Vuetify users (not me, but it's really popular) who have not been able to migrate to Vue 3 yet because Vuetify is incompatible. Hence my [original comment](https://github.com/vuejs/vue/issues/11053#issuecomment-854425611) to "revert [this] before v3 adoption hits critical mass".

**"Comparison with other Frameworks"**
True, React and a couple others do this, but this behavior is an open pain point for React devs (see https://github.com/facebook/react/issues/9230)! Vue gave up what Preact and Riot still have, which is the more intuitive behavior. _What will Vue do if React decides to follow Preact and others?_

**CE**
As a CE and custom HTML author (https://m-docs.org) I understand your points about properties and reflection, but users of CE and custom HTML shouldn't have to be aware of the internal implementation of CE (not applicable to custom HTML). Devs would intuitively expect `<div :attr="false">` to result in `<div>`, as Vue 2, Preact, and Riot do. And authors of components - CE or custom HTML - shouldn't have to do things to ensure their perfectly valid standards-based code is compatible with all the different frameworks. The burden is on the framework to meet developer expectations (re: https://github.com/facebook/react/issues/9230).

As CE inevitably grow in popularity, Vue's behavior will be a sticking point for more and more devs.

I also don't want to push too hard on an emerging pattern because it's mostly me afaik and this isn't even the primary driver behind the issue, but consider:
```html
<m-loader :loading="loading.messages">Getting your messages...</m-loader>
```
That's not a CE. This component is custom HTML implemented with CSS. It may grow up to be a CE one day if JS is ever needed for its features, but today it only requires CSS:
```css
m-loader:before {
  display: inline-block;
  font-family: 'm-icons';
  content: '\e627'; /* "sync" icon */
}

@keyframes m-loader-loading {
  0% { transform: rotate(0deg) matrix(-1, 0, 0, 1, 0, 0) }
  100% { transform: rotate(360deg) matrix(-1, 0, 0, 1, 0, 0) }
}

m-loader[loading]:before {
  animation: m-loader-loading 1.2s linear infinite;
}
```
Vue 3 broke all of my custom HTML that uses boolean attributes. Again, my issue is not a broken app - that's part of the deal when migrating to a major rev - it's the mental gymnastics as I now have to pause and think about attributes, then `null` where necessary, and keep going.

Also, making Vue more CE-friendly (by bringing back the Vue 2 behavior) is **good for Vue adoption** because apps like mine don't get stuck behind huge Vue-centric frameworks like Vuetify that are ironically incompatible with Vue 3 (no diss on Vuetify, but the age of framework-specific design systems is over).

**The deception of "true" and "false" strings**
Special behavior for ARIA and other attributes was introduced based on a mere coincidence that some attributes use "true" and "false" strings and Vue can just happen to produce those exact strings by casting a bool. Not a sound design choice.

In practice, "true" and "false" strings are really no different than "on" and "off", like in the case of `autocomplete`. Vue seems to have been lured in by "true" and "false" as if they were something unique that justified dropping perfectly acceptable and intuitive boolean behavior and introduce a breaking change. 

By this same logic, Vue absolutely should be casting `undefined` as "undefined" to support all these:
- `aria-checked`
- `aria-expanded`
- `aria-hidden`
- `aria-orientation`
- `aria-pressed`
- `aria-selected`

Those six attributes define a "undefined" string in addition to "true" and "false", so it stands to reason then that "undefined" is no more of a "weird edge case" as you claim than "true" and "false" are.

I get the desire to encourage a11y by making ARIA easier (although it's negligible considering `toString()` is all it took), but Vue is now the owner a of faulty pattern.

**The code devs write**
Developers who use ARIA attributes are not or should not (yes "should" is subjective) write code like the examples given. I only see cherry-picked examples.

Let's not forget a few things that put casting bools to strings in the right context for discussing (a narrow context imo):
- First, there are many cases, maybe a majority, where devs simply hard-code them in the template. Vanilla HTML; Vue bindings are not used:
```html
<span aria-haspopup="true">I always have a popup</span>
<div role="timer" aria-atomic="true"></div>
```
- Second, there are many ARIA attributes with "true" and "false", but they don't get used a lot of the time because there's boolean attributes that trigger the same semantics:
```html
<div hidden>
<button disabled>
<input type="checkbox" checked>
<input type="email" required>
<textarea>
and more...

Negates the use of

<div aria-hidden="true">
<button aria-disabled="true">
<input type="checkbox" aria-checked="true">
<input type="email" aria-required="true">
<textarea aria-multiline="true">
```
- Lastly, ARIA attributes default to "false" or "undefined", which makes casting to those strings pointless, which makes _removing_ the attribute in the case of `false` all the more logical (if you really do need explicit "false", possibly for CSS, you set it):
```html
These are the same, so removing is acceptable
<select>
<select aria-disabled="false">
```

Okay, so what we're left with is a very narrow problem space where all 5 of these have to be true:
1. The developer is _dynamically_ setting an attribute (it's easy to forget tons of attributes get hard-coded)
2. And the attribute is one the few "true" or "false" attributes that _don't have a semantic boolean attribute alternative_
3. And the attribute's _presence is required_ with an explicit "false", i.e. removing the attribute and receiving the default false is not acceptable
4. And the developer for some reason wrote code that produces a boolean in the context of an attribute that supports a whole vocabulary of strings, like `aria-haspopup`, `aria-current`, `contenteditable`, and others.
5. And finally, using `toString()` in this severely narrow use case is too hard/unintuitive

That's the scope of this design decision. _Should Vue abandon the power of `false` to solve for that narrow use case?_ The answer was apparently yes, but software doesn't have to be written in stone. 

Bringing back `false` means:
- Familiar boolean attribute behavior is restored (remember, the vast majority of folks are still on Vue 2)
- Better compatibility with CE and custom HTML attributes, which are growing in popularity
- Vue gets rid of these inconsistencies:
  - sometimes `true` and `false` add/remove an attribute and other times it's cast to a string instead of `true` and `false` always adding/removing an attribute.
  - `undefined` _removes_ attributes that define "undefined" (there's 6 of them) but `false` is cast to a string for attributes that define "false"
  - Depending on if the element is a CE and depending on the internal implementation of that CE, `false` may not be cast a string but instead removes the attribute, which produces a different result than what Vue normally would

**Comment by leopiccionia** [comment]
In Lit, `<foo ?bar="${condition}">` renders `<foo bar>` if `condition` is truthy, otherwise renders `<foo>` ([source](https://lit-element.polymer-project.org/guide/templates#bind-properties-to-templated-elements)).

Given that Vue already provides:
* `:foo.prop="value"` (shorthand `.foo="value"`), for forcing a property binding
* `:foo.attr="value"`, for forcing an attribute binding

I propose adding a `v-bind` modifier for toggling boolean attributes in templates.

Long form (naming can be discussed):

```html
<foo :bar.bool="condition"/>
``` 

Short form:

```html
<foo ?bar="condition"/>
```

Rendered content if `condition` is truthy:

```html
<foo bar/>
```

Rendered content if `condition` is falsy:

```html
<foo/>
```

---

A note for CE authors: after [Postel's Law](https://en.wikipedia.org/wiki/Robustness_principle), I'd advise that those predicates should be true (considering that `foo` is boolean, not [enumerated](https://html.spec.whatwg.org/multipage/common-microsyntaxes.html#keywords-and-enumerated-attributes)):
* `<my-el foo="true"/>` === `<my-el foo/>`
* `<my-el foo="false"/>` === `<my-el/>`

My impression is that some frameworks, like [Stencil](https://stenciljs.com/docs/properties), enforces that.

**Comment by jfbrennan** [comment]
I hate to encourage adding more things to learn, more ways of doing the same thing, more bytes to the Vue runtime, and I'm not a fan of proprietary template syntax like `@` and `:` (I wish Vue would embrace standard `onevent` syntax like Riot does), but `?` is an intriguing compromise for sure.
```html
<div ?attr="falsy"> => <div>
<div ?attr="truthy"> => <div attr>
```
I like it. 

From the framework's perspective `?` leaves no doubts what the developer's intentions are. 
From a developer's perspective the expectations for what the framework will do are clear.

This is also a non-breaking opt-in feature for Vue 3. For the 80% who haven't migrated to Vue 3 yet, adding `?` where needed to get the same output they were getting with Vue 2 is an easy no-risk change.

Ship it!

**Comment by jfbrennan** [comment]
@LinusBorg @posva  thoughts on matching Lit's `?` boolean attribute binding syntax as suggested by @leopiccionia?

**Comment by yyx990803** [comment]
Short answer, we will not revert the behavior because of the scale of impact as outlined by @LinusBorg. I think Linus has made the case very clear that it's not possible to revert the behavior without another major version, which we will not consider in the short term. **This decision is final and the team will no longer respond to further requests on reverting the behavior.**

On the other hand, the `?` prefix / `.bool` modifier is something I think could be considered, but should be re-submitted as a separate proposal, using the required RFC format.

**Reply by jfbrennan** [reply]
Oh I know there's no chance of reverting it in Vue 3, it'd have to be a major rev, which is why the discussion was brought here. 

I do understand why the team doesn't want to go through that considering v3 is still relatively new - there's a lot of housekeeping and what not that goes into even small major releases - but the impact is overestimated. Only about 10% of projects use v3 and the use case in question is very narrow (i.e. _dynamically_ setting attrs that need "true" _and_ can't accept the defaults when needing "false" _and_ don't have a semantic boolean).

But no big deal. Lit's `?` is a great solution. Looking forward to it.

**Comment by leopiccionia** [comment]
I'll submit a formal RFC tomorrow for `:attr.bool` and `?attr`, then.
